### ISSUE TITLE
spec negation operator !

### ISSUE THREAD

**Issue by zerothi** [LLNL/spack#23490]
Enable the use of a negation operator in spec dependencies.

### Rationale

It may be viable to never allow a package to use a specific library version for various reasons. Sometimes libraries gets released which then turns out to break API/ABI and a new release is made soon after. But this is not always apparent to end-users or package maintainers.

### Description

From the [spec dependency list](https://spack.readthedocs.io/en/latest/basic_usage.html#specs-dependencies).
For instance to select a version between 1.2 and 1.4, but not 1.3.2 one would do:
```
# current syntax
mpileaks @1.2:1.3.1,1.3.3:1.4 %gcc@4.7.5 +debug -qt target=x86_64 ^callpath @1.1 %gcc@4.7.2
# suggested syntax
mpileaks @1.2:1.4,!1.3.2 %gcc@4.7.5 +debug -qt target=x86_64 ^callpath @1.1 %gcc@4.7.2
```
Since you are already using `^` and `~` it is best if `!` would be used for this (it seems).

### General information

- [x] I have searched the issues of this repo and believe this is not a duplicate

**Comment by alalazo** [comment]
See #23491 I think it solves the issue by marking `1.3.2` as deprecated in the corresponding `package.py`. Otherwise you already point out in the description that this is permitted by the current syntax.

**Comment by zerothi** [comment]
I don't think that is entirely the same usecase. Marking stuff as deprecated requires the `package.py` to be correctly formatted.

It would still be nice to selectively decide version on the command line. E.g. also if one has a suspicion, but no confirmation, about an incompatible version.

While the current syntax could be used, I would argue that the `!version` syntax is much cleaner and shorter without causing confusion.
